### ISSUE TITLE
[JENKINS-21311] Queue item cancellation via REST api works but returns a 404

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-23#3944]
<p>When I call http://<Jenkins_URL>/queue/cancelItem?id=<queueItem> it does cancel the item but it returns a 404. It must return a 200 instead.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=barthelemy">barthelemy</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-21311" target="_blank">Queue item cancellation via REST api works but returns a 404</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/pierrebtz.png" title="pierrebtz's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=pierrebtz">pierrebtz</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Minor
<li><b>component(s)</b>: core
<li><b>label(s)</b>: newbie-friendly, queue, rest, stapler
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2020-03-08T22:18:29+00:00
<li><b>votes</b>: 3
<li><b>watchers</b>: 8
<li><b>imported</b>: 2025-11-23
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>When I call http://&lt;Jenkins_URL&gt;/queue/cancelItem?id=&lt;queueItem&gt; it does cancel the item but it returns a 404. It must return a 200 instead.</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>jenkins 1.546</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-21311?focusedId=213404&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213404">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>There are lots of API endpoints which are not documented, mainly because the creator did not document them. (In this case I am at fault, though I doubt the endpoint I was replacing was documented either.)</p>

<p>So this is simply a bug. Using <tt>HttpResponses.forwardToPreviousPage()</tt> is commonplace in <tt>@RequirePOST</tt> endpoints that are not expected to display anything special where the UI invoker is just something that sends a <tt>POST</tt> via a <tt>form</tt> and does not want the rendered page to change. One easy fix would be to make this method in Stapler check whether there actually is a <tt>Referer</tt>, and if not, just delegate to <tt>ok()</tt> instead.</p>

<p>The proper fix would probably be to think of this as a real API and return 204 No Content, optionally with some body or status message indicating whether anything was actually canceled. It is also necessary to ensure that whatever status code is chosen, <tt>stopButton.jelly</tt> (called from two places with <tt>cancelItem</tt>) will not try to reload the page. From the use of <tt>onclick="new Ajax.Request(this.href); return false"</tt> it looks like this would already be true regardless of the status code.</p>

<p>And it looks like we need a <tt>Queue/_api.jelly</tt> listing the various operations. Currently this gives only generic information.</p>
</pre>
</details>

</ul>
</details>

<p>There are lots of API endpoints which are not documented, mainly because the creator did not document them. (In this case I am at fault, though I doubt the endpoint I was replacing was documented either.)</p>

<p>So this is simply a bug. Using <tt>HttpResponses.forwardToPreviousPage()</tt> is commonplace in <tt>@​RequirePOST</tt> endpoints that are not expected to display anything special where the UI invoker is just something that sends a <tt>POST</tt> via a <tt>form</tt> and does not want the rendered page to change. One easy fix would be to make this method in Stapler check whether there actually is a <tt>Referer</tt>, and if not, just delegate to <tt>ok()</tt> instead.</p>

<p>The proper fix would probably be to think of this as a real API and return 204 No Content, optionally with some body or status message indicating whether anything was actually canceled. It is also necessary to ensure that whatever status code is chosen, <tt>stopButton.jelly</tt> (called from two places with <tt>cancelItem</tt>) will not try to reload the page. From the use of <tt>onclick="new Ajax.Request(this.href); return false"</tt> it looks like this would already be true regardless of the status code.</p>

<p>And it looks like we need a <tt>Queue/_api.jelly</tt> listing the various operations. Currently this gives only generic information.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-21311?focusedId=213435&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213435">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Alright, so it's actually a bug (although I maintain undocumented -&gt; not really API).</p>

<p>Reducing priority as it's essentially cosmetic and can be ignored; maybe just disable following the bogus redirect.</p>
</pre>
</details>

</ul>
</details>

<p>Alright, so it's actually a bug (although I maintain undocumented -> not really API).</p>

<p>Reducing priority as it's essentially cosmetic and can be ignored; maybe just disable following the bogus redirect.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/pierrebtz.png" title="pierrebtz's avatar" /> pierrebtz:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-21311?focusedId=381727&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-381727">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I also lost some time with this return code before seeing this issue, proposed some change to address this <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p>I also lost some time with this return code before seeing this issue, proposed some change to address this <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
